### PR TITLE
document magic constants

### PR DESCRIPTION
**Pull request by robacarp**
For issue #136 

I also added \_\_LINE__ but perhaps that is not needed. As always, comments and suggestions are welcome.

### CODE REVIEW

**Comment by RX14** [comment]
`__LINE__` has a special meaning when used as a default value to a function argument: https://carc.in/#/r/2z97.

I think using a long-form paragraph with examples is going to be a better bet than a table here.

**Comment by mverzilli** [comment]
I agree that a long-form paragraph is better, but the PR adds some valuable info that was missing and maybe we can iterate on this. One thing that I don't like about this is the "Magic" word. Can we move away from that language? Maybe "Predefined constants"?

**Comment by asterite** [comment]
Don't forget about `__END_LINE__`

**Comment by jhass** [comment]
Predefined doesn't cut it for me since they depend on context. How about special purpose constants or pseudo constants (since they're not that constant after all)?

**Comment by mverzilli** [comment]
Context (or contextual) constants sounds great to me!

**Comment by robacarp** [comment]
I rewrote it in paragraph form, and provided an example. I agree this is a better format, good suggestion @RX14. 

@mverzilli, I agree about "Magic" and I think Psuedo Constants conveys the meaning that they're not actually constant better than Contextual Constants. It's also a little less of a mouthful to me. 

Here is the rendered update:

image

**Comment by robacarp** [comment]
Should any of these be documented in this section as well: `ARGF`, `ARGV`, `PROGRAM_NAME`, `STDERR`, `STDIN`, `STDOUT`?

**Comment by asterite** [comment]
@robacarp All of those are "documented" in the API: https://crystal-lang.org/api/0.23.1/toplevel.html

What I mean is that they are there but lack proper documentation.

**Comment by robacarp** [comment]
@asterite yeah, thats where I got the list. I understand the desire to draw the lines well between the std lib and the language reference. Maybe it make sense to have a link in the language reference to the current api docs for those constants?

**Comment by asterite** [comment]
Maybe that makes sense for a tutorial, not for the language reference.

**Comment by RX14** [comment]
I think crystal is really missing a advanced, long, tutorial which brings together the stdlib and language reference.

File: syntax_and_semantics/constants.md [hunk 1]
```diff
@@ -28,3 +28,39 @@ end
 
 TEN #=> 10
 ```
+
+# Psuedo Constants
```
Comment (jhass) [inline syntax_and_semantics/constants.md]: https://en.m.wiktionary.org/wiki/psuedo ☺

Comment (robacarp) [reply syntax_and_semantics/constants.md]: hah, good catch

Comment (robacarp) [reply syntax_and_semantics/constants.md]: I tried to think that I'd typo'd that. "You know how to spell," I told myself. 

And then I found all 6 times I'd typed that. 😅

**Comment by mverzilli** [comment]
Thanks @robacarp!

**Comment by da99** [comment]
Should there be a small mention how these magic constants are also available in macros? Some people coming to Crystal may not have been exposed to a decent macro system and might need reminding this is part of the subset of Crystal available for macros.

**Comment by robacarp** [comment]
@da99 I had the same thought but didn't include anything because [there is a code sample in the macros section](https://crystal-lang.org/docs/syntax_and_semantics/macros.html#constants) which documents this.

**Comment by da99** [comment]
@robacarp Thanks for pointing this out.

**Pull request merged**